src/sage/numerical/pivot_rules_for_simplex_method.py

```diff
@@ -0,0 +1,776 @@
+
+        sage: A = ([1, 1], [3, 1], [-1, -1])
+        sage: b = (1000, 1500, -400)
+        sage: c = (10, 5)
+        sage: P = InteractiveLPProblemStandardForm(A, b, c)
+        sage: def eta(v):
+        ....:     return v.norm()
+        sage: D = P.auxiliary_problem().initial_dictionary() # Phase I
+        sage: D.enter('x0')
+        sage: D.leave('x5')
+        sage: D.update()
+        sage: D.run_simplex_method('NW_rule', eta, [1,1,4])
+        \begin{equation*}
+        ...
+        \end{equation*}
+        Entering: $x_{2}$. Leaving: $x_{0}$.
+        \begin{equation*}
+        ...
+        \end{equation*}
+        sage: d = P.feasible_dictionary(D)
+        sage: d.run_simplex_method('NW_rule', eta, [1,1])
+        \begin{equation*}
+        ...
+        \end{equation*}
+        Entering: $x_{5}$. Leaving: $x_{3}$.
+        \begin{equation*}
+        ...
+        \end{equation*}
+        Entering: $x_{1}$. Leaving: $x_{4}$.
+        \begin{equation*}
+        ...
+        \end{equation*}
+    """
+    def dictionary_pivot(current_dictionary, eta, weight):
+        r"""
+        Sets the entering and leaving variables of `:class:`~sage.numerical.interactive_simplex_method.LPDictionary` as defined by Normalized Weight pivot rule.
+
+        Input:
+        - ``current_dictionary`` -- an instance of `:class:`~sage.numerical.interactive_simplex_method.LPDictionary`
+
+        - ``eta`` -- callable function from ambient space of ``current_dictionary`` to the real numbers.
+
+        - ``weight`` -- object convertible to a sage ``vector`` in the ambient space of current_dictionary.
+
+        EXAMPLES::
+
+            sage: from sage.numerical.pivot_rules_for_simplex_method import SimplexMethodPivot
+            sage: A = ([1, 1], [3, 1])
+            sage: b = (1000, 1500)
+            sage: c = (10, 5)
+            sage: P = InteractiveLPProblemStandardForm(A, b, c)
+            sage: D = P.initial_dictionary()
+            sage: D.entering()
+
+            sage: D.leaving()
+
+            sage: def eta(v): # use 2 norm as normalization
+            ....:     return v.norm()
+            sage: pivot = SimplexMethodPivot("NW_rule")
+            sage: pivot(D, False, eta, [1,1])
+            sage: D.entering()
+            x1
+            sage: D.leaving()
+            x4
+
+            Inputs are assumed to be correct for the current problem::
+            sage: D.update()
+            sage: pivot(D, False, eta, 1)
+            Traceback (most recent call last):
+            ...
+            TypeError: 'sage.rings.integer.Integer' object is not iterable
+            sage: eta = [1,2,1]
+            sage: pivot(D, False, eta, [1,1])
+            Traceback (most recent call last):
+            ...
+            TypeError: 'list' object is not callable
+        """
+        eta = eta
+        weight = vector(weight)
+        if current_dictionary.entering() is None:
```

Comment:
Not sure if silently doing nothing if the entering variable is already set is a good interface.
Consider other options: Just override what was selected before; or raise an error.

Reply:
I think a change like this would introduce behavior that isn't expected in a 2 phase LP. For example, in the initial step of the auxiliary problem, the entering variable is not determined by the pivot rule (being x0), but the leaving variable should be (at least this is what I would think the expected behavior should be in this case). By overriding we'd enter potentially shouldn't be entered at this point or we'd raise and error when we shouldn't be raising an error.

Reply:
this being said, there is a change that should be made in the interactive simplex method that should reflect using the pivot rule at all steps.